js/ui/grid_core/ui.grid_core.keyboard_navigation.js

```diff
@@ -697,24 +743,63 @@ const KeyboardNavigationController = core.ViewController.inherit({
 
     _navigateNextCell: function($event, keyCode) {
         const $cell = this._getNextCell(keyCode);
+        const directionCode = this._getDirectionCodeByKey(keyCode);
+        const isCellValid = $cell && this._isCellValid($cell);
+        const result = isCellValid ? this._arrowKeysHandlerFocusCell($event, $cell, directionCode) : false;
+        return result;
+    },
 
-        if($cell && this._isCellValid($cell)) {
-            return this._arrowKeysHandlerFocusCell($event, $cell, true);
+    _arrowKeysHandlerFocusCell: function($event, $nextCell, direction) {
+        const isVerticalDirection = direction === 'prevRow' || direction === 'nextRow';
+        if(this._isVirtualColumnRender() && !isVerticalDirection && !this._focusVirtualCell($nextCell, direction)) {
+            return false;
         }
 
-        return false;
-    },
+        const args = this._fireFocusChangingEvents($event, $nextCell, isVerticalDirection, true);
 
-    _arrowKeysHandlerFocusCell: function($event, $cell, upDown) {
-        const args = this._fireFocusChangingEvents($event, $cell, upDown, true);
-        $cell = args.$newCellElement;
-        if(!args.cancel && this._isCellValid($cell)) {
-            this._focus($cell, !args.isHighlighted);
+        $nextCell = args.$newCellElement;
+
+        if(!args.cancel && this._isCellValid($nextCell)) {
+            this._focus($nextCell, !args.isHighlighted);
             return true;
         }
         return false;
     },
 
+    _focusVirtualCell: function($nextCell, direction) {
+        const nextCellPosition = this._getCellPosition($nextCell);
+        const cellModel = this._getCellModel(
+            nextCellPosition.rowIndex,
+            nextCellPosition.columnIndex
+        );
+        const column = cellModel && cellModel.column;
```

Comment:
The column field is used in cellModel. So, Why _getCellModel method is created. The _getColumnByIndex method will be simpler.

Reply:
_focusVirtualCell is not required anymore. I removed it.